src/indexer.rs

```diff
@@ -132,32 +134,80 @@ async fn sync_state_to_db(
+                sync_state_to_db(abi_item.name.to_lowercase(), csv_writer, db_writers).await;
             }
 
-            if !indexer_config.postgres.apply_indexes_before_sync && !reached_head {
-                println!(
-                    "applying indexes for {}, may take a little while...",
-                    abi_item.name
-                );
-                generate_event_table_indexes(postgres_db, abi_item, &abi_item.name)
-                    .await
-                    .unwrap();
+            // The block below is very specific to postgres
+            // This is some scaffolding to allow this to work for now
+            // TODO: move index creation to postgres implementation + make part of interface for datasource
+            if let Some(postgres_conf) = &indexer_config.postgres {
+                if !postgres_conf.apply_indexes_before_sync && !reached_head {
+                    println!(
+                        "applying indexes for {}, may take a little while...",
+                        abi_item.name
+                    );
+
+                    // Apply to postgres client from db writer (db client) list
+                    for datasource in db_writers {
+                        if let Some(postgres_db) =
+                            datasource.as_any().downcast_ref::<PostgresClient>()
+                        {
+                            generate_event_table_indexes(postgres_db, abi_item, &abi_item.name)
+                                .await
+                                .unwrap();
+                        }
+                    }
+                }
             }
         }
     }
 }
 
+///
+///  This is a factory to declare the various supported datasources
+///  For now this will check existence of each type block to add to implementation (postgres or
+///  bigquery) - but should consider a more generic implementation in the future
+///
+///  # Arguments
+///
+///  * indexer_config: the full configuration    
+pub async fn init_datasource_writers(
+    indexer_config: &IndexerConfig,
+) -> Vec<Box<dyn DatasourceWritable>> {
+    // Return a list / vector of datasource writers
+    let mut writers: Vec<Box<dyn DatasourceWritable>> = Vec::new();
+
+    // Init postgres client (if exists)
+    if let Some(postgres_conf) = &indexer_config.postgres {
+        let postgres_db_client =
+            init_postgres_db(&postgres_conf, &indexer_config.event_mappings, false)
+                .await
+                .expect("Failed to initialize Postgres client");
+        writers.push(Box::new(postgres_db_client));
+    }
+
+    // Init GCP bigquery client (if exists)
+    if let Some(bigquery_conf) = &indexer_config.gcp_bigquery {
+        let bigquery_db_client =
+            init_gcp_bigquery_db(&bigquery_conf, &indexer_config.event_mappings)
+                .await
+                .expect("Failed to initialize bigquery client");
+        writers.push(Box::new(bigquery_db_client));
+    }
+
+    writers
```

Comment:
you think we should throw here if none of the above exist? i cant imagine them using postgres and bigtable together or is that what you have imagined?

Reply:
sounds good, added a check to the bottom of the method.  we should have at least one indexer configured, otherwise wouldn't make sense for the sync to proceed.

Reply:
i had envisioned that we could allow for multiple writers - conceivably if someone wanted to write to both postgres as well as gcp (or any other datasource for that matter) - they all operate on the generated csv file.  `init_datasource_writers` returns a vector of `DatasourceWritable`s.  the functions `sync_state_to_db` as well as `sync_all_states_to_db` loops through multiple `DatasourceWritable`s calling its `write_data` method).

Reply:
if it works we can call it a feature then haha